collectors/proc.plugin/proc_pagetypeinfo.c

```diff
@@ -0,0 +1,344 @@
+                "mem"
+                , "pagetype_global"
+                , NULL
+                , "pagetype"
+                , NULL
+                , "System orders available"
+                , "B"
+                , PLUGIN_PROC_NAME
+                , PLUGIN_PROC_MODULE_PAGETYPEINFO_NAME
+                , NETDATA_CHART_PRIO_SYSTEM_MEMFRAG
+                , update_every
+                , RRDSET_TYPE_STACKED
+            );
+            for (o = 0; o < MAX_PAGETYPE_ORDER; o++) {
+                char id[3+1];
+                snprintfz(id, 3, "%lu", o);
+
+                char name[20+1];
+                dim_name(name, o);
+
+                systemorders[o].rd = rrddim_add(st_order, id, name, PAGE_SIZE, 1, RRD_ALGORITHM_ABSOLUTE);
+                rrddim_set_by_pointer(st_order, systemorders[o].rd, systemorders[o].size);
+            }
+        }
+
+
+        // Per-Numa Node & Zone & Type (full detail). Only if sum(line) > 0
+        st_nodezonetype = callocz(cnt_zones, sizeof(RRDSET));
+        for (p = 0; p < pagelines_cnt; p++) {
+            pgl = &pagelines[p];
+
+            // Skip invalid, refused or empty pagelines if not explicitely requested
+            if (!pgl
+                || do_detail == CONFIG_BOOLEAN_NO
+                || (do_detail == CONFIG_BOOLEAN_AUTO && pageline_total_count(pgl) == 0))
+                continue;
+
+            // "pagetype Node" + NUMA-NodeId + ZoneName + TypeName
+            char setid[13+1+2+1+MAX_ZONETYPE_NAME+1+MAX_PAGETYPE_NAME+1];
+            snprintfz(setid, 13+1+2+1+MAX_ZONETYPE_NAME+1+MAX_PAGETYPE_NAME, "pagetype_Node%d_%s_%s", pgl->node, pgl->zone, pgl->type);
+
+            // Skip explicitely refused charts
+            if (simple_pattern_matches(filter_types, setid))
+                continue;
+
+            // "Node" + NUMA-NodeID + ZoneName + TypeName
+            char setname[4+1+MAX_ZONETYPE_NAME+1+MAX_PAGETYPE_NAME +1];
+            snprintfz(setname, MAX_ZONETYPE_NAME + MAX_PAGETYPE_NAME, "Node %d %s %s",
+                pgl->node, pgl->zone, pgl->type);
+
+            st_nodezonetype[p] = rrdset_create_localhost(
+                    "mem"
+                    , setid
+                    , NULL
+                    , "pagetype"
+                    , NULL
+                    , setname
+                    , "B"
+                    , PLUGIN_PROC_NAME
+                    , PLUGIN_PROC_MODULE_PAGETYPEINFO_NAME
+                    , NETDATA_CHART_PRIO_MEM_PAGEFRAG + p
+                    , update_every
+                    , RRDSET_TYPE_STACKED
+            );
+            for (o = 0; o < MAX_PAGETYPE_ORDER; o++) {
+                char dimid[3+1];
+                snprintfz(dimid, 3, "%lu", o);
+                char dimname[20+1];
+                dim_name(dimname, o);
+
+                pgl->rd[o] = rrddim_add(st_nodezonetype[p], dimid, dimname, PAGE_SIZE, 1, RRD_ALGORITHM_ABSOLUTE);
+            }
+        }
+    }
+
+    if(unlikely(!cnt_nodes)) {
```

Comment:
Check it before chart initialization and swap `cnt_nodes` with `pagelines_cnt`.

Reply:
You've changed the variable, but you also shouldn't try to initialize charts if you can't get data from the file. Move this check and the error message to L181.

Reply:
Between these lines
```
        // Init the RRD graphs

        // Per-Order: sum of all node, zone, type Grouped by order
```

Reply:
Added more checks for values and cleanup. If we can't get the data from file (pageline_cnt == 0) we return with error.
If we didn't had any valid line for `pagetypeinfo_line_valid` return too with error.

Reply:
Hm. I don't understand what's the difference between `p` and `pagelines_cnt`?

And why do you need
```
    if(unlikely(!pagelines_cnt)) {
        error("PLUGIN: PROC_PAGETPEINFO: Cannot find the number of NUMA Nodes in %s", PLUGIN_PROC_MODULE_PAGETYPEINFO_NAME);
        return 1;
    }
```
after charts initialization?

Reply:
I've made the validity check sooner, against `pagelines_cnt` directly and removed this dead code. It was for when I planned to do per-numa node summaries, but it isn't useful anymore.